roachtest: sqlsmith/setup=empty/setting=no-mutations failed

[(roachtest).sqlsmith/setup=empty/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1874488&tab=buildLog) on [master@29c0efdcc5edb5d100449a093b25df107f1df2d6](https://github.com/cockroachdb/cockroach/commits/29c0efdcc5edb5d100449a093b25df107f1df2d6):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200414-1874488/sqlsmith/setup=empty/setting=no-mutations/run_1
	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:753: error: pq: internal error: top-level relational expression cannot have outer columns: (2)
		stmt:
		WITH
			with_5031 (col_27826)
				AS (
					SELECT
						*
					FROM
						(
							VALUES
								(8779.999671960051834:::DECIMAL),
								((-3.86254023746085817E+20):::DECIMAL),
								(3.545032605710859015E+37:::DECIMAL),
								(7.814952987538403048E+36:::DECIMAL)
						)
							AS tab_16132 (col_27826)
				)
		SELECT
			min(tab_16134.col_27829::TIME) OVER (PARTITION BY tab_16134.col_27829)::TIME AS col_27833
		FROM
			(
				VALUES
					(
						'9 years 4 mons 340 days 24:28:12.393926':::INTERVAL,
						'[{"(;n9": {"b": [{"a": [[], 0.8002157518786387]}]}, "^6D": {"1=9]": 2.5313843630958095}, "c": true, "foobar": [null]}, {}, []]':::JSONB
					),
					(NULL, (NULL::JSONB - 'ocY)}yDo':::STRING::STRING)::JSONB),
					(
						'47 years 6 mons 892 days 18:55:40.069485':::INTERVAL,
						'{"b": true, "baz": [[[[[]], {}, 0.7033812934795172, null], true], []]}':::JSONB
					)
			)
				AS tab_16133 (col_27827, col_27828),
			(VALUES ('10:25:00.627138':::TIME), ('18:58:37.355154':::TIME), ('05:47:05.085575':::TIME))
				AS tab_16134 (col_27829),
			(VALUES ('\xb24c':::BYTES)) AS tab_16135 (col_27830),
			with_5031 AS cte_ref_1444
		GROUP BY
			tab_16134.col_27829
		HAVING
			not_ilike_escape(e'8a\x0f$':::STRING::STRING, (SELECT e'~\x0b\x02zMC[':::STRING AS col_27832 FROM (VALUES ((1802136919109728215:::INT8::INT8 * '30 years 1 mon 184 days 17:34:47.280666':::INTERVAL::INTERVAL)::INTERVAL), (tab_16133.col_27827), ('61 years 3 mons 404 days 14:06:37.400319':::INTERVAL), (tab_16133.col_27827), (tab_16133.col_27827)) AS tab_16136 (col_27831) WHERE true ORDER BY tab_16136.col_27831 DESC, tab_16136.col_27831 LIMIT 1:::INT8)::STRING, '$l':::STRING::STRING)::BOOL
		LIMIT
			83:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=empty/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=1874488&tab=artifacts#/sqlsmith/setup=empty/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dempty%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/opt/optbuilder/subquery.go
+++ b/pkg/sql/opt/optbuilder/subquery.go
@@ -217,7 +217,9 @@ func (s *subquery) buildSubquery(desiredTypes []*types.T) {
 	defer func() { s.scope.builder.subquery = outer }()
 	s.scope.builder.subquery = s
 
-	outScope := s.scope.builder.buildStmt(s.Subquery.Select, desiredTypes, s.scope)
+	// We must push() here so that the columns in s.scope are correctly identified
+	// as outer columns.
+	outScope := s.scope.builder.buildStmt(s.Subquery.Select, desiredTypes, s.scope.push())
 	ord := outScope.ordering
 
 	// Treat the subquery result as an anonymous data source (i.e. column names

--- a/pkg/sql/opt/optbuilder/testdata/subquery
+++ b/pkg/sql/opt/optbuilder/testdata/subquery
@@ -2225,3 +2225,9 @@ build
 SELECT 3::decimal IN (SELECT 1::int)
 ----
 error (22023): unsupported comparison operator: <decimal> IN <tuple{int}>
+
+# Regression test for #47467 - we weren't correctly identifying a.x as an outer column.
+build
+SELECT 1 FROM (VALUES (1)) AS a(x) HAVING (SELECT true FROM (VALUES (a.x)) AS b(y))
+----
+error (42803): subquery uses ungrouped column "x" from outer query

